core/src/main/java/com/linecorp/armeria/internal/common/LeakTracingRequestContextStorage.java

```diff
@@ -54,7 +54,7 @@ final class LeakTracingRequestContextStorage implements RequestContextStorage {
     public <T extends RequestContext> T push(RequestContext toPush) {
         requireNonNull(toPush, "toPush");
         if (sampler.isSampled(toPush)) {
-            return delegate.push(wrapRequestContext(toPush));
+            return delegate.push(wrapRequestContext(toPush.unwrapAll()));
```

Comment:
Can you check the following test and try adding `assertThat(ClientRequestContext.current()).isSameAs(clientCtx);`?

https://github.com/line/armeria/blob/cae3b6da373101c0976e2d00f1f3cbb706d07e00/it/trace-context-leak/src/test/java/com/linecorp/armeria/internal/common/TraceRequestContextLeakTest.java#L77

I'm not sure I understand why we're pushing the root context 😅

Reply:
> Can you check the following test and try adding

Added. 😉 

> we're pushing the root context

We're not pushing the root context but the original context with just wrapped by `TraceableRequestContext` only once. Previously we are pushing the context something like
`TraceableRequestContext(TraceableRequestContext(TraceableRequestContext(... originalCtx)))`

Reply:
But wouldn't `originalCtx` also `unwrap`?  Anyways, happy with the changes if the commit passes 😄

Reply:
> But wouldn't originalCtx also unwrap?

I also worry about this case that a non-tracing wrapper is unwrapped.
How about `unwrap()` once when `toPush` is instance of `TraceableClientRequestContext` and `TraceableServiceRequestContext`?

Reply:
> But wouldn't originalCtx also unwrap?

Forgot to leave a comment about this. Had a chat with @jrhee17 and he misunderstood the behavior. 😄 

> I also worry about this case where a non-tracing wrapper is unwrapped.

I thought there's no such case but let me add a logic that prevents the situation.

Reply:
> there's no such case 

Yeah. It would be a rare case that I've never seen so far. 😅 (possible somehow though)

Reply:
Updated. 😆